scripts/monitor/visualization/resource_monitoring.py

```diff
@@ -0,0 +1,596 @@
+        sharded_tasks = dict([(k, v) for k, v in tasks.items() if v != 0])
+        # unsharded_tasks = dict(filter(lambda elem: elem[1] == 0, tasks.items()))  # pycharm false positive?
+        # sharded_tasks = dict(filter(lambda elem: elem[1] != 0, tasks.items()))
+        for task in unsharded_tasks:
+            self.plot_aggregate(task, -1)
+        for task in sharded_tasks:
+            self.save_bokeh(task)
+
+    def combine(self, *args):  # UNTESTED: NO IDEA IF THIS WORKS
+        """
+        Combines multiple class instances (with their workflow data)
+        Meant for class instances with the same workflow
+        :param args: other class instances to combine- for data on multiple workflows
+        :return: None
+        """
+        Frames = list(self.meta_runtime)
+        for class_instance in args:
+            if (class_instance.meta_runtime.columns == self.meta_runtime.columns)[0]:
+                #
+                # NOTE TO SELF: this is stupid. It doesn't even work. Just check the workflow IDs.
+                # But do it later, when you fix the function for real
+                # Combining the datasets like this won't really work, figure something else out later
+                #
+                Frames.append(class_instance.meta_runtime)
+            else:
+                raise Exception("Not the same workflow, data cannot be combined")
+        self.meta_runtime = pd.concat(Frames)
+
+    def num_task_shards(self) -> Dict[str, int]:
+        """
+        Returns dict with task names as key and shard count as value
+        :return: dict of task:shard_count
+        """
+
+        task_and_shard_idx = self.meta_runtime.loc[:, ['runtime_task_call_name', 'runtime_shard']]
+        res = dict()
+        for task in task_and_shard_idx['runtime_task_call_name'].unique():
+            res[task] = 1 + max(task_and_shard_idx
+                                [task_and_shard_idx['runtime_task_call_name'] == task]['runtime_shard'])
+        return res
+
+    def disk_requested_for_task(self, task: str) -> pd.DataFrame:
+        """
+        Disk requested for a task (only successful runs, if sharded display all shards)
+        :param task: task name
+        :return: dataframe of disk requested for task
+        """
+        columns_to_select = ["runtime_task_call_name", "runtime_shard", "runtime_disk_total_gb",
+                             "meta_disk_total_gb", "meta_execution_status"]
+        renaming_scheme = {"runtime_task_call_name": "task",
+                           "runtime_shard": "shard_idx"}
+        df = self.meta_runtime.loc[:, columns_to_select].rename(columns=renaming_scheme)
+        df = df[df["meta_execution_status"] == "Done"]
+        if task in df['task'].unique():
+            return df[df["task"] == task].sort_values(by=['shard_idx'])
+        else:
+            raise KeyError("Task name does not exist")
+
+    def mem_requested_for_task(self, task: str) -> pd.DataFrame:
```

Comment:
`mem_requested_for_task` and `disk_requested_for_task` share common code, and so that leads me to thinking:
what about having a class `TaskResource`

```python

class Resource:

    def __init__(self, ...):
        self.cpu_platform = ...
        self.cpu_count = ...
        self.memory = ...
        self.disk = ...
        self.gpu_platform = ...
        self.gpu_count = ...


class TaskResource:

    def __init__(self, ...):
        self.name = ...
        self.shard = -1 if unshared else some_vale
        self.inputs = dict(...)

        self.resource_requested = Resource(...)
        self.resource_allocated = Resource(...)
        
        self.resource_at_time = [Resource(...), Resource(...), ]
        
```

Basically, this models the concept of a task from the perspective of its resource usage.

Obviously this is going to take a significant effort to restructure parts of your code.